### PR TITLE
[BACKLOG-18489] -  Added method to create extensional representation of a filter.

### PR DESCRIPTION
**Pull request by graimundo**
@pentaho/millenniumfalcon please review

### CODE REVIEW

File: impl/client/src/main/javascript/web/pentaho/data/filter/abstract.js [hunk 1]
```diff
@@ -429,6 +429,59 @@ define([
         return result;
       },
 
+      /**
+       * Gets the [extensional]{@link https://en.wikipedia.org/wiki/Extensional_definition} representation of a [Filter]{@link pentaho.data.filter.spec.IFilter}.
+       *
+       * @param {!pentaho.data.AbstractTable} dataTable
+       * @param {!String[]} keyColumnNames
```
Comment (dcleao) [inline impl/client/src/main/javascript/web/pentaho/data/filter/abstract.js]: --> `{string[]}`  (already non-null)

File: impl/client/src/main/javascript/web/pentaho/data/filter/abstract.js [hunk 2]
```diff
@@ -429,6 +429,59 @@ define([
...
+       *
+       * @return {!pentaho.data.filter.Or|!pentaho.data.filter.False}
+       */
+      toExtensional: function(dataTable, keyColumnNames) {
+        var filteredData = dataTable.filter(this);
+        var columnNameToIdx = getColumnIdToIndexMap( filteredData, keyColumnNames );
```
Comment (dcleao) [inline impl/client/src/main/javascript/web/pentaho/data/filter/abstract.js]: Spaces `( .. )`

File: impl/client/src/main/javascript/web/pentaho/data/filter/abstract.js [hunk 3]
```diff
@@ -429,6 +429,59 @@ define([
...
+        var orOperands = [];
+        var numRows = filteredData.getNumberOfRows();
+        for(var rowIdx = 0; rowIdx < numRows; rowIdx++) {
+          var andOperands = keyColumnNames.map(function(keyColumnName) {
+            var cellValue = filteredData.getValue(rowIdx, columnNameToIdx[keyColumnName]);
+            var columnType = filteredData.getColumnType(columnNameToIdx[keyColumnName]);
+            return new __filter.IsEqual({property: keyColumnName, value: {_: columnType, v: cellValue}});
+          });
+
+          if(andOperands.length > 0) {
```
Comment (dcleao) [inline impl/client/src/main/javascript/web/pentaho/data/filter/abstract.js]: The only way to have 0 length is for keyColumnNames to have 0 length...

File: impl/client/src/main/javascript/web/pentaho/data/filter/abstract.js [hunk 4]
```diff
@@ -429,6 +429,59 @@ define([
...
+
+        // no conditions were specified
+        return __filter.False.instance;
+
+
+        /**
+         * Gets an object where the keys are the column names and the values the respective column index.
+         *
+         * @param {!pentaho.data.ITable} dataTable
+         * @param {!string[]} columnNames
```
Comment (dcleao) [inline impl/client/src/main/javascript/web/pentaho/data/filter/abstract.js]: `{string[]}`

File: impl/client/src/main/javascript/web/pentaho/data/filter/abstract.js [hunk 5]
```diff
@@ -429,6 +429,59 @@ define([
         return result;
       },
 
+      /**
+       * Gets the [extensional]{@link https://en.wikipedia.org/wiki/Extensional_definition} representation of a [Filter]{@link pentaho.data.filter.spec.IFilter}.
+       *
+       * @param {!pentaho.data.AbstractTable} dataTable
```
Comment (dcleao) [inline impl/client/src/main/javascript/web/pentaho/data/filter/abstract.js]: Parameters need descriptions (at least in public methods:
`@param {!pentaho.data.ITable} dataTable - The data table on which to evaluate the extension.`

Comment (dcleao) [reply impl/client/src/main/javascript/web/pentaho/data/filter/abstract.js]: Should not reveal the `AbstractTable` class, if possible

File: impl/client/src/main/javascript/web/pentaho/data/filter/abstract.js [hunk 6]
```diff
@@ -429,6 +429,59 @@ define([
...
+        return __filter.False.instance;
+
+
+        /**
+         * Gets an object where the keys are the column names and the values the respective column index.
+         *
+         * @param {!pentaho.data.ITable} dataTable
+         * @param {!string[]} columnNames
+         *
+         * @returns {!{}} key
```
Comment (dcleao) [inline impl/client/src/main/javascript/web/pentaho/data/filter/abstract.js]: `@returns {!Object.<string, number>}`

**Comment by graimundo** [comment]
@nantunes @dcleao updated PR addressing review comments

**Pull request merged**